ocs-ktc/src/main/kotlin/org/ostelco/prime/ocs/core/OnlineCharging.kt

```diff
@@ -30,91 +34,111 @@ object OnlineCharging : OcsAsyncRequestConsumer {
         val msisdn = request.msisdn
 
         if (msisdn != null) {
-
             CoroutineScope(Dispatchers.Default).launch {
 
                 val response = CreditControlAnswerInfo.newBuilder()
                         .setRequestId(request.requestId)
                         .setMsisdn(msisdn)
                         .setResultCode(ResultCode.DIAMETER_SUCCESS)
 
-                if (request.msccCount > 0) {
-                    val mscc = request.getMscc(0)
-                    val requested = mscc?.requested?.totalOctets ?: 0
-                    val used = mscc?.used?.totalOctets ?: 0
-
-                    val responseMscc = MultipleServiceCreditControl
-                            .newBuilder(mscc)
-                            .setValidityTime(86400)
-
-
-                    storage.consume(msisdn, used, requested) { storeResult ->
-                        storeResult.fold(
-                            {
-                                // TODO martin : Should we handle all errors as NotFoundError?
-                                response.resultCode = ResultCode.DIAMETER_USER_UNKNOWN
-                                synchronized(OnlineCharging) {
-                                    returnCreditControlAnswer(response.build())
-                                }
-                            },
-                            {
-                                consumptionResult ->
-
-                                val grantedTotalOctets = if (mscc.reportingReason != ReportingReason.FINAL
-                                        && mscc.requested.totalOctets > 0) {
-
-                                    if (consumptionResult.granted < mscc.requested.totalOctets) {
-                                        responseMscc.finalUnitIndication = FinalUnitIndication.newBuilder()
-                                                .setFinalUnitAction(FinalUnitAction.TERMINATE)
-                                                .setIsSet(true)
-                                                .build()
+                val doneSignal = CountDownLatch(request.msccList.size)
+
+                request.msccList.forEach { mscc ->
```

Comment:
forEach is not async.  So, we might not need CountDownLatch.

Reply:
ok, I added it since the storage access is async